### ISSUE TITLE
Want to Fine-tuning BiReFNet and Restoring Validation Loss

### ISSUE THREAD

**Issue by Soumya2564** [ZhengPeng7/BiRefNet#211]
Hello @ZhengPeng7 ,

I plan to fine-tune your BiReFNet model on my custom dataset to improve its performance on specific image types where it currently struggles, using your BiReFNet model(BiRefNet-general-epoch_244.pth) as the base (DIS5K-TR, DIS-TEs, DUTS-TR_TE, HRSOD-TR_TE, UHRSD-TR_TE, HRS10K-TR_TE, TR-P3M-10k, TE-P3M-500-NP, TE-P3M-500-P, [TR-humans]). To ensure we follow the best approach, we would appreciate clarification on the following:-

1) What is the recommended strategy for fine-tuning BiReFNet to improve performance on challenging image types?

2) Should we combine your original training dataset(DIS, HRSOD, and COD) with our 40,000 new images which is doesn't perform and train on the combined dataset, or should we fine-tune only on our new dataset?

3) How many epochs should we train the model for during fine-tuning?

4) Should we adjust the learning rate during fine-tuning? If so, what learning rate do you recommend?

5) Are there any additional recommendations or best practices for fine-tuning BiReFNet on a domain-specific dataset?

6) Lastly, how can we add validation loss during training, as it have been removed in your updated code?

**Comment by ZhengPeng7** [comment]
Here are my suggestions for reference. You need to make the adjustment depending on your real cases.
1. If you have more specific things, we can talk about some potential strategies. Otherwise, basic operations would be enough.
2. Use only the custom dataset, except that some existing datasets are very very similar to your cases.
3. It depends on the size of the fine-tuning data. Suppose it has 500 images, 50 epochs would be enough.
4. The default learning rate should be okay for most cases, as I used before.
5. Some strategies on the data loading, things like hard negative mining, may help.
6. Add the inference + eval part after each `train_epoch` in the `train.py`.

**Comment by Soumya2564** [comment]
Hi @ZhengPeng7 ,
I have a dataset of over 46,000 human images, which I used to train my model up to 230 epochs. However, the final trained model (epoch_230.pth) does not perform well on a separate set of 20,000 human images.

Can I fine-tune my model using this 20,000-image dataset, or is this dataset too large for fine-tuning?

**Comment by ZhengPeng7** [comment]
Yeah, that's quite a large number. It's better to train the model on it from scratch.

**Comment by Soumya2564** [comment]
Hi @ZhengPeng7 
Thanks for your previous reply. I have another question:-

Can I fine-tune the model with 5,000 images? If not, what is the maximum dataset size recommended for fine-tuning?

Previously, you mentioned that fine-tuning with 500 images is possible. In that case, can I fine-tune the model 20 times using different sets of 500 images each?

**Comment by ZhengPeng7** [comment]
If the data used in pre-training is too different from your custom, it might do some harm. If very similar, that'll always be good.
And your custom data is already a large number. It's better to train from scratch.

**Comment by Soumya2564** [comment]
Hi @ZhengPeng7 ,
Thank you for all your help!

I saw in a previous response where you mentioned that it's possible to train the model at a lower resolution. I’d like to clarify: can we train the model from scratch using 256×256 or 512×512 resolution images?

If so, which pretrained weights would you recommend as the starting point for training at lower resolution?
a) reso_dynamic-general-epoch_174
b) general-epoch_244

Also, if we train at a lower resolution, will it affect the final output quality? Will there be any noticeable difference in quality compared to the original model trained at higher resolution?

**Comment by ZhengPeng7** [comment]
Sure. Weights pre-trained in higher resolutions tend to exhibit better performance, even when the resolutions used for fine-tuning are lower (as observed in my experiments).

Regarding choosing one of them, I can assure you. They should work similarly.

**Comment by Soumya2564** [comment]
Hi @ZhengPeng7  ,

I came across your General-Dynamic demo output on Hugging Face and was really impressed by the results. Could you please guide me on where I can access or download the General-Dynamic model?

**Comment by ZhengPeng7** [comment]
Both in my [HuggingFace](https://huggingface.co/ZhengPeng7/BiRefNet_dynamic), the [Github Release](https://github.com/ZhengPeng7/BiRefNet/releases), and the [google drive](https://drive.google.com/drive/u/0/folders/1s2Xe0cjq-2ctnJBR24563yMSCOu4CcxM).
